File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class WordDictionary {
    class Node{
        Node[] links;
        boolean flag;

        public Node(){
            links = new Node[26];
            flag = false;
        }
    }

    private Node root;

    public WordDictionary() {
        root = new Node();
    }
    
    public void addWord(String word) {
        Node curr = root;
        for(char c:word.toCharArray()){
            int idx = c - 'a';
            if(curr.links[idx]==null){
                curr.links[idx] = new Node();
            }
            curr = curr.links[idx];
        }
        curr.flag = true;
    }
    
    public boolean search(String word) {
        return find(0, word.toCharArray(), root);
    }

    private boolean find(int idx, char[] charArray, Node curr){
        if(idx==charArray.length) return curr.flag;
        char c = charArray[idx];

        if(c=='.'){
            for(Node child: curr.links){
                if(child!=null && 
                    find(idx+1, charArray, child)){
                        return true;
                }
            }
            return false;
        }
        else{
            int i = c - 'a';
            if(curr.links[i]==null) return false;
            return find(idx+1, charArray, curr.links[i]);
        }
    }
}


/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class WordDictionary(object):
    class TrieNode:
        def __init__(self):
            self.children = {}
            self.endOfWord = False

    def __init__(self):
        self.root = self.TrieNode()

    def addWord(self, word):
        """
        :type word: str
        :rtype: None
        """
        curr = self.root
        for c in word:
            if c not in curr.children:
                curr.children[c] = self.TrieNode()
            curr = curr.children[c]
        curr.endOfWord = True
        
    def search(self, word):
        """
        :type word: str
        :rtype: bool
        """
        n = len(word)

        def dfs(idx, curr):
            if idx == n:
                return curr.endOfWord

            find = word[idx]

            if find == '.':
                for child in curr.children.values():
                    if dfs(idx + 1, child):
                        return True
                return False
            else:
                if find in curr.children:
                    return dfs(idx+1, curr.children[find])
                return False

        return dfs(0, self.root)

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)